robot/impl/local_robot.go

```diff
@@ -784,17 +865,43 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
 		if err != nil {
 			return
 		}
-		if len(r.getWeakDependencyMatchers(conf.API, conf.Model)) == 0 {
+
+		// Return early if resource has neither weak nor optional dependencies.
+		if len(r.getWeakDependencyMatchers(conf.API, conf.Model)) == 0 &&
+			len(conf.ImplicitOptionalDependsOn) == 0 {
 			return
 		}
-		r.Logger().CDebugw(ctx, "handling weak update for resource", "resource", resName)
+		r.Logger().CDebugw(ctx, "handling weak/optional update for resource", "resource", resName)
 		deps, err := r.getDependencies(resName, resNode)
 		if err != nil {
-			r.Logger().CErrorw(ctx, "failed to get dependencies during weak dependencies update; skipping", "resource", resName, "error", err)
+			r.Logger().CErrorw(
+				ctx,
+				"failed to get dependencies during weak/optional dependencies update; skipping",
+				"resource", resName,
+				"error", err,
+			)
 			return
 		}
-		if err := res.Reconfigure(ctx, deps, conf); err != nil {
-			r.Logger().CErrorw(ctx, "failed to reconfigure resource during weak dependencies update", "resource", resName, "error", err)
+
+		// Use the module manager to reconfigure the resource if it's a modular resource. This
+		// would be a modular resource that has optional dependencies.
+		isModular := r.manager.moduleManager.Provides(conf)
```

Comment:
If we were to _just_ call `Reconfigure` instead of routing reconfigures to the module manager when necessary, we would hit a no-op `Reconfigure` for known APIs and a "this resource does not implement reconfigure" for foreign resources (custom APIs). The testing I was doing with the `optionaldepsmodule` was not even working 😮‍💨 . Confirmed with @aldenh-viam offline that this conditional to call `moduleManager.ReconfigureResource` gets optional dependencies down to Python modules. Will also write a test that leverages modular implicit optional dependencies in addition to builtin modular implicit optional dependencies.

Reply:
we ought to refactor this block out (at some point) so we don't run into this issue next time 😬

Reply:
I might push back there a bit. This code is so fragile/opaque that, in some senses, it's kind of nice to see what's explicitly happening in each code path (with some custom comments for both the path here in local_robot.go and the other in resource_manager.go) and avoid jumping around with go-to-definition in your IDE. I would also _not_ expect this exact code pattern to be used a third time. If you feel strongly that you'd like it to be a function, I can still do it, though.